### PR TITLE
Autodetect service name for spring boot applications

### PR DESCRIPTION
**Pull request by amarziali**
# What Does This Do

Instruments spring boot application environment to use the value of `${spring.application.name}` as a default service name when a value of it has not been provided thought usual ways (`DD_SERVICE` env variable or equivalents).

The instrumentation will only apply for a standalone spring boot application (running as a jar). 
It won't apply to applications packaged as a war that are deployed on application servers. This because an application server may host multiple applications/services hence the standard approach (use the servlet context) will still apply.

Please also note that this instrumentation introduces a breaking changes in terms of default service name calculation. Despite we can put that behind a feature flag, I think that it can be acceptable to have this for standalone spring boot application since, in my mind, it's uncommon to have a dd_service undefined (or not having tagged a container) for this kind of application. Having it on by default allows to have less option to provide to instrument and can turned it off by disabling the instrumentation itself in case of unwanted behaviour.

## Why the changeset is so big?

The spring environment bootstraps after the tracer itself hence there is the need to have a way to lazily set or mutate the service name when it is not inferred and defaults to `unnamed-java-app` or the jar name.
Today we are setting the service name as an immutable value on the Config object. This value is initialised in a very early stage and it's used in almost every product needing to tag data (tracing, profiling, debugging, ciapp, telemetry, statsd, jmx, ...). 

In order to make that value mutable, we introduced a `ServiceNaming` object that holds service name value and is responsible for updating it if possible (i.e. the user did not provide a value for `DD_SERVICE`).

# Motivation

To ease up the onboarding, we'd like to automatically generate service names for spring boot applications especially when the application is not yet instrumented. This will allow the datadog agent to surface meaningful application names on the service catalogue.
However, in order to be coherent, the same defaults should apply when the application is instrumented. Also, those values needs to be stable for the service map and also in order to fetch the same configuration if defined remotely (remote config)

# Additional Notes

I also provided a complementary JMH benchmark for the statsD client. In fact, in order to keep the service name mutable, I had to remove it from all the precomputed values that depends to it.  In particular, the statsD client may be one of the most sensitive part and we previously put particular care of how static tags were precomputed.

Even if the service name is no more part of that packed tags, I tried to make it as optimized as I could.
The following results may reassure around performance impact on that part

```
Benchmark                                                 (tagParam)  Mode  Cnt  Score   Error  Units
TagCombinerBenchmark.benchDynamicServiceName                          avgt    2  3.418          ns/op
TagCombinerBenchmark.benchDynamicServiceName  some:thing,other:value  avgt    2  5.779          ns/op
TagCombinerBenchmark.benchStaticServiceName                           avgt    2  2.634          ns/op
TagCombinerBenchmark.benchStaticServiceName   some:thing,other:value  avgt    2  4.760          ns/op
```

Notes: currently reviewing load tests to see which part is most uncovered concerning this changeset

Jira ticket: [AIT-9541]

[AIT-9541]: https://datadoghq.atlassian.net/browse/AIT-9541?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by jpbempel — COMMENTED** [review]
Overall I am uncomfortable with this change.
We are dropping on the floor here a lot of assumptions regarding `DD_SERVICE`
The possibility that the service can change name on the fly and subject of spring boot instrumentation is running and when means that, until it happens, we have another service name that all data and module that rely on and later will be change!

File: communication/src/jmh/java/datadog/communication/monitor/TagCombinerBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.infra.Blackhole;
+
+@State(Scope.Benchmark)
+@Warmup(iterations = 1, time = 30, timeUnit = SECONDS)
+@Measurement(iterations = 2, time = 30, timeUnit = SECONDS)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(MICROSECONDS)
```
[the remark this answers is not in the document]
Comment (amarziali) [reply communication/src/jmh/java/datadog/communication/monitor/TagCombinerBenchmark.java:28]: I updated the measurement unit. The avg difference is within 1nanos and let me think that the overhead might be considered negligible comparing to the rest

File: internal-api/src/main/java/datadog/trace/api/naming/ServiceNaming.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+  public ServiceNaming(final CharSequence initialName, final boolean mutable) {
+    this.original = UTF8BytesString.create(initialName);
+    this.mutable = mutable;
+    doUpdate(original);
+  }
+
+  // this is just to ensure that updates are not running wildly. There is no consistency on the read
+  // to avoid contentions.
+  private synchronized void doUpdate(@Nonnull final UTF8BytesString name) {
```
[the remark this answers is not in the document]
Comment (amarziali) [reply internal-api/src/main/java/datadog/trace/api/naming/ServiceNaming.java]: yes it's true. I simplified as suggested.

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/ConfigurationUpdater.java [hunk 3]
```diff
@@ -147,7 +148,7 @@ private Configuration applyFiltersAndAddExceptionProbes(
         .addAllowList(configuration.getAllowList())
         .addDenyList(configuration.getDenyList())
         .setSampling(configuration.getSampling())
-        .setService(configuration.getService())
+        .setService(serviceNaming.getSanitizedName().toString())
```
[the remark this answers is not in the document]
Comment (amarziali) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/agent/ConfigurationUpdater.java]: thank you I wasn't sure about. I changed accordingly

**Comment by amarziali** [comment]
> Overall I am uncomfortable with this change. We are dropping on the floor here a lot of assumptions regarding `DD_SERVICE` The possibility that the service can change name on the fly and subject of spring boot instrumentation is running and when means that, until it happens, we have another service name that all data and module that rely on and later will be change!

Hi @jpbempel thanks for the review and I understand the concerns. The major issue I consider we have with the service name is that it's used for different processes (and perhaps different meanings) and the result is that we couple tracing with other products with this key while we could have another independent key to identify the observed application.  For spring boot one improvement we'd like to do is to have this service name autodetected and that should be used as well for the remote configuration and the service map (that's sourced from stats/spans).  The Configuration is initialized quite at very first instants the agent bootstraps. Unless we limit the feature to taking the spring application name from sysprops, env or args without introspecting the packaged properties (that's in my mind does not fully reflect the most common way people is deploying applications) I do not see an easy way to avoid mutating the service name. As alternative deferring processes that depends from the service name seems not viable as well since there are no guarantee of the "when" it will be stable. If we assume that this service name change will happen once and at very early stage (that's true for spring boot) we can probably accept that some telemetry/metric might be sent with another service name. One improvement I might do is verify that, if the spring boot environment is refreshed, the service name change is not triggered again. Otherwise I should freeze it to change only once

**Comment by amarziali** [comment]
As discussed separately, will close in favour of a simplified version

**Pull request closed without merging**